File: algorithms/M/Minimum Cost to Hire K Workers/Minimum Cost to Hire K Workers.java
class node implements Comparable<node>{
    int q;
    int w;
    double wpq;
    node(int qual, int wage){
        q = qual;
        w = wage;
        wpq = (double)w/q;
    }
    
    public int compareTo(node o){
        return Double.compare(this.wpq, o.wpq);
    }
}
class Solution {
    
    public double mincostToHireWorkers(int[] quality, int[] wage, int k) {
        
        ArrayList<node> arr = new ArrayList<>();
        //main constant value of importance is wage/quality
        for(int i = 0; i< wage.length; i++) {
            arr.add(new node(quality[i], wage[i]));
        }
        
        Collections.sort(arr);
        double res = Double.MAX_VALUE;
        PriorityQueue<Integer> maxheap = new PriorityQueue<>(Collections.reverseOrder());
        int sum = 0;
        for(node n: arr) {
            
            maxheap.add(n.q);
            sum = sum+n.q;
            
            if(maxheap.size()>k){
                sum -= maxheap.poll();
            }
            
            if(maxheap.size() == k) {
                res = Math.min(res,sum*n.wpq);
            }
        }
        
        return res;
        
    }
}

File: algorithms/M/Minimum Cost to Hire K Workers/Minimum Cost to Hire K Workers.js
/**
 * @param {number[]} quality
 * @param {number[]} wage
 * @param {number} k
 * @return {number}
 */
var mincostToHireWorkers = function(quality, wage, k) {
  let minCost = Number.MAX_SAFE_INTEGER;
  let offer, accepted, ratio;
  for(let i=0; i<quality.length; i++) {
    ratio = wage[i]/quality[i];
    accepted = [];
    for(let j=0; j<quality.length; j++) {
      offer = quality[j] * ratio;
      if(offer >= wage[j]) accepted.push(offer);
    } 
    if(accepted.length < k) continue;
    minCost = Math.min(minCost, accepted.sort((a,b) => a-b).splice(0, k).reduce((acc, cur) => acc+cur, 0));
  }
  return minCost;
}
